### ISSUE TITLE
test_allreduce_coalesced_stress is flaky

### ISSUE THREAD

**Issue by mrshenli** [pytorch/pytorch#25427]
Passed: https://circleci.com/gh/pytorch/pytorch/2624121
Failed: https://circleci.com/gh/pytorch/pytorch/2624119

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera

**Comment by mrshenli** [comment]
test disabled in #25419

**Comment by jfc4050** [comment]
ok so i think i have some idea what is going on. I can't reproduce this locally and I only really have the CI check output to go off, so I have no way of knowing if my proposed changes will fix the issue but let me know if you want me to make a PR

so this is the error message from CI
```
Aug 29 19:19:12 test_allreduce_coalesced_stress (__main__.ProcessGroupGlooTest) ... ERROR:root:Caught exception: le() received an invalid combination of arguments - got (unicode), but expected one of:
Aug 29 19:19:12  * (Tensor other)
Aug 29 19:19:12       didn't match because some of the arguments have invalid types: (!unicode!)
Aug 29 19:19:12  * (Number other)
Aug 29 19:19:12       didn't match because some of the arguments have invalid types: (!unicode!)
Aug 29 19:19:12 , exiting process with exit code: 10
```

the call to `assertEqual` in `_test_allreduce_coalesced_stress`
```python
            self.assertEqual(
                2 * [torch.Tensor([(i * self.world_size) + (self.world_size * (self.world_size - 1) / 2)])],
                inputs[i],
                "Mismatch in interation {}".format(i)
            )
```
I noticed that in #25334 `None` was added between the expected output and the Mismatch* string for the `assertEqual` calls in each of the existing stress tests. This wasn't added to the `_test_allreduce_coalesced_stress` `assertEqual` call, probably because it hadn't been merged for very long at the time

the signature and first few lines of `assertEqual` looks like this
```python
    def assertEqual(self, x, y, prec=None, message='', allow_inf=False):
        if isinstance(prec, str) and message == '':
            message = prec
            prec = None
        if prec is None:
            prec = self.precision
```
and later what i think is the offending call.
```python
self.assertLessEqual(max_err, prec, message)
```

what i think is happening is that the Mismatch* string is being converted to unicode somehow, making `isinstance(prec, str)` False and resulting in a unicode string being passed to `self.assertLessEqual`

my proposed change is just passing the Mismatch* string in as a kwarg like so
```python
            self.assertEqual(
                2 * [torch.Tensor([(i * self.world_size) + (self.world_size * (self.world_size - 1) / 2)])],
                inputs[i],
                message="Mismatch in interation {}".format(i)
            )
```
or just adding `None` as was done in #25334

**Comment by mrshenli** [comment]
@jfc4050 thanks for the detailed analysis. This looks reasonable to me. One nit is that, would it be more clear if we use `message="Some error message"` instead of setting `prec`?

@pritamdamania87 Do you know why the string message is converted to unicode? Curious, how did you come up with the adding `None` solution?

**Comment by mrshenli** [comment]
@jfc4050 could you please also take a look at #25465? Somehow `ProcessGroupGloo` thinks `t1.cuda()` and `t2.cuda()` are using different device types.

**Comment by jfc4050** [comment]
> @jfc4050 thanks for the detailed analysis. This looks reasonable to me. One nit is that, would it be more clear if we use `message="Some error message"` instead of setting `prec`?

yes, that would be my preferred change, seems more clear and would be less likely to be broken in the future.

**Comment by jfc4050** [comment]
> @jfc4050 could you please also take a look at #25465? Somehow `ProcessGroupGloo` thinks `t1.cuda()` and `t2.cuda()` are using different device types.

I have to leave in about 30mins, but i'll take a look and make a comment in #25465 if I find anything between now and then. Be back on for about an hour at noon

**Comment by pietern** [comment]
@jfc4050 @pritamdamania87 Any idea why the message kwarg issue resulted in test flakiness? I expected this to either fail all the time or not at all, instead of being intermittent.

**Comment by pietern** [comment]
Closing this because the fix in #25557 is landing.

**Comment by jfc4050** [comment]
> @jfc4050 @pritamdamania87 Any idea why the message kwarg issue resulted in test flakiness? I expected this to either fail all the time or not at all, instead of being intermittent.

sorry for the late response,
it doesn't seem like the unicode messages should have flaky behavior to me either. I'm not too familiar with how python2 handles unicode but I'll look around a little and see if I can come up with an answer
